### PR TITLE
signed webhooks + automatically assign to activejobs

### PR DESCRIPTION
**Pull request by Skaelv**
- add an option in the generator to automatically sign and verify webhooks in the lib
- webhook controller has only one method which calls a job following a convention : 
- 'conversations.user.replied' webhook calls ConversationUserRepliedJob.perform

### CODE REVIEW

**Comment by Skaelv** [comment]
@choran

File: README.md [hunk 1]
```diff
@@ -115,13 +114,15 @@ You can add your webhooks subscriptions to the `IntercomApp.config`. Subscriptio
 IntercomApp.configure do |config|
   config.webhooks = [
     {topics: ['users'], url: 'https://my-intercom-app.com/webhooks/users'},
-    {topics: ['conversation.user.created', 'conversation.user.replied'], url: 'https://my-intercom-app.com/webhooks/conversations'}
+    {topics: ['conversation.user.created', 'conversation.user.replied'], url: 'https://my-intercom-app.com/webhooks/conversations'},
```
Comment (mariokostelac) [inline README.md]: my-intercom-app.com should be probably loaded from the config itself

File: app/controllers/intercom_app/webhooks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    end
+
+    private
+
+    def webhook_params
+      params.except(:controller, :action, :type)
+    end
+
+    def webhook_job_klass
+      "#{webhook_type.sub('.','_').classify}Job".safe_constantize or raise IntercomApp::MissingWebhookJobError
```
Comment (mariokostelac) [inline app/controllers/intercom_app/webhooks_controller.rb]: maybe better MissingWebhookJobDefinitionError?

File: app/controllers/intercom_app/webhooks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+module IntercomApp
+  class WebhooksController < ApplicationController
+    skip_before_action :verify_authenticity_token
+
+    include IntercomApp::WebhookVerification
+
+    class IntercomApp::MissingWebhookJobError < StandardError; end
+
+    def receive
+      params.try(:permit!)
```
Comment (mariokostelac) [inline app/controllers/intercom_app/webhooks_controller.rb]: what is this?

File: lib/generators/intercom_app/home_controller/templates/home_controller.rb [hunk 4]
```diff
@@ -1,8 +1,6 @@
-class HomeController < ApplicationController
-  around_filter :intercom_session
+class HomeController < IntercomApp::AuthenticatedController
 
   def index
-    @users_count = @intercom_client.users.all.count()
-    @users_names = @intercom_client.users.all.map(&:name)
+    @users_names = @intercom_client.users.find_all(:type=>'users', :per_page => 10, :page => 1, :order => "desc", :sort => "created_at").map(&:name)
```
Comment (mariokostelac) [inline lib/generators/intercom_app/home_controller/templates/home_controller.rb]: can we isolate this to a separate PR?

File: lib/generators/intercom_app/install/install_generator.rb [hunk 5]
```diff
@@ -35,11 +40,20 @@ def create_intercom_session_repository_initializer
...
       end
 
-
       def mount_engine
         route "mount IntercomApp::Engine, at: '/'"
       end
 
+      private
+
+      def generate_signature
```
Comment (mariokostelac) [inline lib/generators/intercom_app/install/install_generator.rb]: generate_hub_secret is a better name i guess

Comment (mariokostelac) [inline lib/generators/intercom_app/install/install_generator.rb]: it's not really clear what signature you're generating. you're not signing webhooks, you're receiving them

File: lib/intercom-app/utils.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+require 'digest/sha1'
+
+module IntercomApp
+  module Utils
+
+    def random_string
+      (0...20).map { (65 + rand(26)).chr }.join
+    end
+
+    def webhook_signature
```
Comment (mariokostelac) [inline lib/intercom-app/utils.rb]: can we rename this to something more meaningful like `generate_new_hub_secret`?

File: lib/intercom-app/webhook_verification.rb [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  module WebhookVerification
+    extend ActiveSupport::Concern
+
+    included do
+      before_action :verify_request
+    end
+
+    private
+    def verify_request
+      p request.headers
```
Comment (mariokostelac) [inline lib/intercom-app/webhook_verification.rb]: this should probably be omitted :)

File: lib/intercom-app/webhooks_manager.rb [hunk 8]
```diff
@@ -41,6 +40,7 @@ def is_required_webhook?(webhook)
     end
 
     def create_webhook(attributes)
```
Comment (mariokostelac) [inline lib/intercom-app/webhooks_manager.rb]: this should probably be named `create_webhook_subscription`

File: lib/intercom-app/webhooks_manager.rb [hunk 9]
```diff
@@ -51,5 +51,10 @@ def webhook_exists?(topics)
     def current_webhooks
       @current_webhooks ||= intercom_client.subscriptions.all.index_by(&:topics)
     end
+
+    def sign_webhook(attributes)
```
Comment (mariokostelac) [inline lib/intercom-app/webhooks_manager.rb]: and this one `add_hub_secret`

**Comment by mariokostelac** [comment]
Naming is odd, but feel free to 🚢 after fixing that.

**Pull request merged**